### ISSUE TITLE
When using md-waves button stays dimmed

### ISSUE THREAD

**Issue by MaximBalaganskiy** [aurelia-ui-toolkits/aurelia-materialize-bridge#168]
When I put `md-waves="color: light;"` instead of `waves-effect waves-light` and click the button it stays dimmed after a "wave" until I move the mouse off the button.

This does not happen with the original waves classes.

**Comment by Thanood** [comment]
Interesting..

It doesn't seem to do that in the catalog app. To rule out any problems with my monitor or anything like that: can you confirm that this also happens in the [catalog app](http://aurelia-ui-toolkits.github.io/demo-materialize/#/samples/waves/colors) (second button uses light waves).

I think what this bridge does different compared to Materialize is that it calls ```Waves.attach(this.element);``` on each element while Materialize calls ```Waves.displayEffect();``` on ```DOMContentLoaded```. 

I also don't see it in [this gist](https://gist.run/?id=895ab84a029eb49f460babef88bf498d). But, as said, that may be my monitor.

Please also tell me what kind of button you use.
```<a md-button>```, ```<button md-button>```, anything without ```md-button``` or anything else? There may be visual differences and I'm just trying the wrong things here. :smile:

**Comment by MaximBalaganskiy** [comment]
Indeed, it's something in my setup - waves overlay each other. If i click a button several times it gets lighter/darker each time....

**Comment by Thanood** [comment]
Can you recreate you setup to the point of being reproducable in a gist.run? You can f.i. fork the one linked above for that.

**Comment by MaximBalaganskiy** [comment]
I'm using requirejs to load aurelia. I guess they are too big for gist :(

**Comment by Thanood** [comment]
Not the whole project. :smile: 
Only the parts necessary to reproduce this issue.

If it's not possible to reproduce, it may as well be due to how requirejs loads modules.
To clarify this:
Usually the ```DOMContentLoaded``` event is triggered a (relatively) long time before Aurelia attaches any views which leads to postpone initialization of dynamically added elements. The bridge was created to simplify this with custom attributes and elements.

If the requirejs setup somehow manages to attach views before ```DOMContentLoaded``` triggers that might lead to Waves being initialized twice.

But honestly, that is just a guess.
Having said that, I unfortunately don't know where to start looking if I cannot reproduce this behaviour.

I can try to do both initialization with a Materialize Codepen, though. I'll see what comes out of it and post it here for reference.

**Comment by MaximBalaganskiy** [comment]
I took my bundles to create a gist and fail to reproduce it. I guess this one can be closed - it's something in my project.

**Comment by MaximBalaganskiy** [comment]
Solved it! With your help @Thanood! It was double initialization indeed.
I had some leftover code which worked when I user materialize directly.

**Comment by Thanood** [comment]
Great! :+1:

**Comment by Ganbin** [comment]
Actually I run into the same issue.

I have read on the official materialize-css ([Issue#2667](https://github.com/Dogfalo/materialize/issues/2667)) that this can be caused by a double initialization. Which is not my case.

I'm using the Aurelia CLI tools and it require a modified version of materialize-css.

I have read the help for installing the dependencies wihtin an Aurelia CLI project (https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/issues/205)

But I can't figure out why this behaviour happend in my project.

**Comment by Ganbin** [comment]
Just to add a details, the waves disapear when I mouve in and out the cursor of my mouse. So it look like it is a problem of event which doesn't detach the effect after some miliseconds...

**Comment by Ganbin** [comment]
And if I add only the class `class="waves-effect"` it work normally.

So this issue only happend with the `md-waves` attribute.

**Comment by MaximBalaganskiy** [comment]
@Ganbin put a breakpoint in materialize.js file where waves is initialized and check the code is not called twice

**Comment by Ganbin** [comment]
@MaximBalaganskiy It is only called once.

**Comment by Thanood** [comment]
> And if I add only the class class="waves-effect" it work normally.

That shouldn't be the case at all when Materialize is only initialized once. It works like this (in short):
* Materialize is initialized
* calls `Waves.displayEffect()` [here](https://github.com/Dogfalo/materialize/blob/4e5804ea59da9eef074f195bb9281f9d6e77f6df/js/waves.js#L335) on `DOMContentLoaded` event.

Now here's the thing: DOMContentLoaded is raised *long before*  aurelia takes control - if it is only fired once. It's been the case with webpack that this event has been fired twice. That may be a reason here, too. Although I think it's unlikely..

So all these initializations (including Waves) in theory take place before any Aurelia related things start. I say "in theory" because there is nothing to initialize - unless you add `class="waves-effect"` onto some element.

Then after some time, `md-waves` gets `attached()` calling `Waves.attach(this.element)`.
If you have `class="waves-effect"` *and* `md-waves` on the same element, it would get initialized twice if it's in the DOM from the start. In this case you get the initialization from both `Waves.displayEffect()` and after that `Waves.attach()`.

Waves event handlers are triggered by `mouseup` and `mouseleave` so I guess the first `Effect.hide` is called at `mouseup` while the second effect is only called at `mouseleave`

It really sounds like Waves is initialized twice in some way - maybe not Materialize but Waves.

**Comment by Ganbin** [comment]
Ok now I see.

It is initialize once but like you said, the button element have both `md-waves` and the `class="waves-effect` but this sound like it is add by the aurelia bridge because in my code I have only this : 

`<button md-waves md-button><i class="left material-icons">mode_edit</i>I'm a basic button</button>`

So the initialization is only one but the `showEffect` function is fired twice, But that's sound strange because it look like it is the aurelia-bridge that add dynamically this class like we can see [here (waves.js)](https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/blob/master/dist/amd/waves/waves.js#L81).

But if I comment this out, then the effect is no more bind to the button. So I'm still investigating in why the effect is fired twice.

**Comment by Ganbin** [comment]
My button element is bind two time to the same event listener, one time [here](https://github.com/Dogfalo/materialize/blob/4e5804ea59da9eef074f195bb9281f9d6e77f6df/js/waves.js#L329) and one time [here](https://github.com/Dogfalo/materialize/blob/4e5804ea59da9eef074f195bb9281f9d6e77f6df/js/waves.js#L308)

So I don't know why it this bind atomatically by the `attach` function. actually I don't understand good the comments of the dev for this function.

**Comment by Ganbin** [comment]
So the [`showEffect`](https://github.com/Dogfalo/materialize/blob/4e5804ea59da9eef074f195bb9281f9d6e77f6df/js/waves.js#L278) function is fired twice but the [`attach`](https://github.com/Dogfalo/materialize/blob/4e5804ea59da9eef074f195bb9281f9d6e77f6df/js/waves.js#L329) is apply only once.

**Comment by Ullfis** [comment]
It seems like its not needed to attach the element when the materialize-css is used as amd module. I have only tested this with aurelia-cli. It may be other and better ways to solve this.

[replace this](https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/blob/master/dist/amd/waves/waves.js#L93) with:

```javascript
if (!(typeof require === "function" && typeof require.specified === "function" && require.specified('waves'))) {
  Waves.attach(this.element);
}
```

[is-it-possible-to-detect-if-a-script-is-being-loaded-as-a-requirejs-module](http://stackoverflow.com/questions/18085022/is-it-possible-to-detect-if-a-script-is-being-loaded-as-a-requirejs-module)

**Comment by Ganbin** [comment]
@Ullfis Thanks but it seems not to solve my problem. But if I remove the `!` in your condition it make it to works...

**Comment by Ganbin** [comment]
It seems that to add the `waves-effect` to the element to make the waves effect to work. So this Waves.attach is not necessary and will make trigger twice the event.

**Comment by Thanood** [comment]
@Ganbin To recap: you're using CLI which uses requirejs. But if you remove the `!` then the condition becomes true when `require` is defined. So `Waves.attach()` is called in your case. But that should not be the case because *then* the effect is attached twice..

I'm confused. :smile:

**Comment by Ganbin** [comment]
@Thanood If I keep the `!` then the condition is `true`. This is what happens in my code.

But what is the purpose of doing `Waves.attach(element)` if adding the `waves-effect` class is enough to activate the waves effect?

**Comment by Thanood** [comment]
@Ganbin this here

> It seems like its not needed to attach the element when the materialize-css is used as amd module.

was news to me.

But this applies to AMD only (and I still wonder how it works then).
`Waves-attach()` is generally needed when you add elements after `DOMContentLoaded` which would normally be the case in an Aurelia based application (especially when you do routing).

The reason is when you add elements *after* `DOMContentLoaded` then the Waves "trigger" which starts with `DOMContentLoaded` cannot add the effect.

Now if that's totally different when using AMD (or the Materialize fork, didn't look at it this way) then I'll have to check this out.

**Comment by Ganbin** [comment]
I see that in my project, I don't use the https://github.com/ccloli/materialize but https://github.com/Dogfalo/materialize.git like recommended in this comment (https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/issues/205#issuecomment-236288489)

My aurelia.json look like this :

    {
      "name": "jquery",
      "path": "../node_modules/jquery/dist",
      "main": "jquery.min"
    },
    {
      "name": "materialize-css",
      "path": "../node_modules/materialize-css/dist",
      "main": "js/materialize",
      "deps": [
        "jquery"
      ],
      "resources": [
        "css/materialize.css"
      ]
    },
    {
      "name": "aurelia-materialize-bridge",
      "path": "../node_modules/aurelia-materialize-bridge/dist/amd",
      "main": "index",
      "resources": [
        "./**/*.{html}",
        "./**/*.{css}"
      ]
    }

and for now I didn't get any other strange behaviour except this waves effect.

**Comment by Thanood** [comment]
Okay, confirmed without `!`:

```javascript
// Waves.attach(this.element);
      if ((typeof window.require === "function" && typeof window.require.specified === "function" && window.require.specified('waves'))) {
        Waves.attach(this.element);
      }
```

![aurelia-cli-waves](https://cloud.githubusercontent.com/assets/677826/18122032/d4635bee-6f67-11e6-90a9-ab94b3079553.gif)

**Comment by Thanood** [comment]
What about a build condition? Since this is **only** needed when AMD is used (I still don't understand why...), this would be quite reliable when only included in the AMD build.

**Comment by Thanood** [comment]
Interesting..
When I do this in `index.html`:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>Aurelia</title>
    <script>
      document.addEventListener('DOMContentLoaded', function() {
        console.log('DOMContentLoaded');
        console.log(Waves); // <-- this
    }, false);
    </script>
  </head>

  <body aurelia-app="main">
    <script src="scripts/vendor-bundle.js" data-main="aurelia-bootstrapper"></script>
  </body>
</html>
```

Waves is defined and since it's at `DOMContentLoaded` it is also initialized. The button `waves-effect` class, set by `md-waves` is still added later. There's still something I don't see. :smile:

**Comment by Thanood** [comment]
Some enlightenment from @MaximBalaganskiy :smile: 

:point_up: [7. September 2016 07:24](https://gitter.im/aurelia-ui-toolkits/aurelia-materialize-bridge?at=57cfa49d8965597230483327)

> when bridge does not depend on materialize directly
> attached can be called before materialize is finished loading
...
> the JSPM config uses amd which is async

**Comment by JoschaMetze** [comment]
Hi, I'm jumping the train late, but I have the same problem in a different setup. 
I'm using webpack (with typescript).
What I don't quite understand yet (and maybe someone can help me) is why do we have two event handlers for mouse down in this case?
Waves.displayEffect adds the handler on the body-element and Waves.attach adds the handler on the button element (or whatever element the attach is called on).
So when the button is clicked both event handler fire and the wave effect is applied twice (and doesn't disappear completely on mouseup or mouseleave.

You can reproduce the effect on http://zrvis.hochfrequenz.org/ , click the FAB button multiple times to see the effect.

**Comment by Thanood** [comment]
Hi @JoschaMetze,

first we thought this was an issue with JSPM using `amd` which works async.. But I guess that's wrong. @MaximBalaganskiy pointed out that this is probably because the Materialize bundle finishes after the bridge is initialized and everything is attached. In this case `Waves.displayEffect` is called *after*  everything is attached.

This main.js seems to work with JSPM, it uses SystemJS:
https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/issues/263#issuecomment-245863046

As you may see, it's used to ensure the bridge is initialized after `document.ready`. I don't know how this would be solved with webpack. Do you have an idea?

I can't answer your question about why we need two event handlers directly, you would have to ask this the Waves team.. But `Waves.displayEffect` is supposed to be called when everything in the DOM is ready. In this case an event handler on `body` seems more efficient to me.

`Waves.attach` is actually used for single elements in original Waves. That's why we use this in our components.

btw: I'm confused why the catalog app doesn't have that problem (bundled JSPM app). In fact the current workaround removed all Waves from the catalog app.

**Comment by JoschaMetze** [comment]
The catalog app is a good point. Did you try running it from the github site(http://aurelia-ui-toolkits.github.io/demo-materialize/#/samples/waves), or the local version? Because on the github site the wave effect doesn't work at all for me.

**Comment by Thanood** [comment]
Right, I only noticed that today. But it's only remotely related.. I will try and switch the jspm configuration of the bridge to SystemJS for that app and revert the workaround. 

Is it possible in webpack to declare a dependency on materialize in its configuration?

**Comment by JoschaMetze** [comment]
I've tried to load materialize in webpack's aurelia-bootstrap module and the bridge in a dependent module but that didn't work either.
@niieani, maybe you can help us with this issue, as you are the webpack expert?

**Comment by Thanood** [comment]
My goal with this was:
Let Materialize load completely so it can attach its `document.ready` handler, then create a promise which is also resolved in `document.ready` where aurelia is bootstrapped.

This should make sure that Materialize doesn't initialize after Aurelia runs because you attach a second callback to the promise which Materialize already used. This way it should always run second.

I just don't know how to do that in webpack. Or if it's necessary at all to create pseudo-dependencies.

**Comment by JoschaMetze** [comment]
I think you reached the goal with your modification to main.ts, but it doesn't help with my waves :).
Materialize is loaded before any aurelia attached calls, so even DOMContentLoaded is called before any attached calls.
So the order is:

1. materialize is initialized the callback to DOMContentLoaded is added as an event handler
2. aurelia-materialize-bridge is loaded via webpack
3. via your main.ts promise the document.ready is called and DOMContentLoaded is triggered
4. materialize calls Waves.displayEffect() (and adds an eventHandler on mousedown)
5. aurelia builds the page and the md-waves attached attribute is called
6. Waves.attached is called in materialize (and adds an eventHandler on mousedown)
7. on clicking the button both event handlers are triggered.

**Comment by Thanood** [comment]
Well, that's the point.. :)

> 4. materialize calls Waves.displayEffect() (and adds an eventHandler on mousedown)
> 5. aurelia builds the page and the md-waves attached attribute is called

If it's like this then there's no problem because before 5 starts, there is nothing "waveable" in the DOM. `displayEffect` should have no ... effect. 😄 

Of course if you can't guarantee that materialize is loaded before adding Aurelia's bootstrap to the `document.ready` promise, it may be added as the second consumer.

I guess I'll have to try this in webpack..

**Comment by Thanood** [comment]
Oh, I see you mean the click listener takes over anyway..

**Comment by JoschaMetze** [comment]
That's what's dazzling me most. Just by thinking about how materialize waves should work, your description should work. But looking at the code I can't really grasp how this works at all with jspm or anything else.
`displayEffect` doesn't wrap anything as no `.waves-effect` is present in the dom. But the event handler is set.
`attached` (in bridge) does add the classes and `Waves.attach` adds the second event handler.
So two event handlers call `showEffect`(both setting the target to the button, as it originates from the mousedown). 
`showEffect` then calls `Effect.show` thus adding the waves elements twice (once for each event handler)

**Comment by Thanood** [comment]
Out of ideas for now.
Maybe it "clicks" when I see it myself in webpack. Sometimes it's just like that. 😃 

What if you remove `Waves.attach` from the bridge source in node_modules. Does it work? And does it work *every time* then? If not (every time) then this would at least be the same problem as with `amd`.

**Comment by JoschaMetze** [comment]
Uncommenting the `Waves.attach` works in 20 of 20 cases I tried. So it is a solution for me, dunno if you can get something generic from that.

**Comment by Thanood** [comment]
@JoschaMetze I've downloaded a fresh esnexr-webpack skeleton, applied the installation instructions and changed the button in `welcome.html` to use waves. It doesn't stay dimmed.. 😕 

I kept bootstrap, though. Maybe this impacts loading order, I'll try without.

**Comment by Thanood** [comment]
@JoschaMetze
Update on this:
I could just reproduce this with a webpack-skeleton version beta-2.0.0 but the current skeleton 1.0.0 still doesn't have this problem for me.

**Comment by JoschaMetze** [comment]
@Thanood,

could you upload your skeleton somewhere, so I can easily check for the differences, or is it the plain vanilla skeleton (download or master)?

**Comment by Thanood** [comment]
I don't have it here (at work). It's the vanilla esnext-webpack-skeleton with installation instructions applied from our catalog app ([here](http://aurelia-ui-toolkits.github.io/demo-materialize/#/installation) so you don't need to look it up).

I'll recreate it when I have some time.
If not, I'll upload my original when I come home.

**Comment by JoschaMetze** [comment]
Ahhh, I should have read your instructions before, now the difference is clear. You are adding the materialize-library to the app-bundle (not adding it to any other bundle and excluding from app). This works but has the disadvantage that the app bundle is larger and has to be downloaded every time the app itself changes. So to reproduce my error add the materialize-css lib to the bootstrap bundle (and remove the import in main.js/main.ts)

**Comment by Thanood** [comment]
I will try that, thank you..

So my conclusion (or "not-so-wild guess" in other words) is that in my case Materialize is in the same bundle as the app, thus not causing any async problems with packages (which is also still a guess after all).

I may be able to reproduce this using your suggestion but atm I'm lost with webpack. I will try to find a solution where the bridge can "depend" on Materialize.. In the sense that Materialize is always loaded first.

**Comment by Thanood** [comment]
Okay, I understand that what you describe happens [here](https://github.com/aurelia/skeleton-navigation/blob/master/skeleton-esnext-webpack/webpack.config.js#L60). But then I noticed how limited my webpack knowledge really is. 😕 

So I need to catch up on this. Especially on how to exclude the bridge from the app bundle.
In the Aurelia docs there is a chapter about bundling with webpack. There is also a paragraph about async bundles - which I suppose happens here.
Also loading with `System.import` seems to be possible.

I would really love to understand all this. 😃 

The easy (cheap? cheating? idk) way seems to be loading the materialize bundle with System.import and then continue with an equivalent version of the main.js here: https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/issues/263#issuecomment-245863046

Maybe some "chunk-magic" is the real answer to this.
Will take some time to catch up on webpack.

**Comment by Thanood** [comment]
With aurelia-cli this could be solved by using separate bundles.
I'm not sure why but the effect is that Waves are **not** inititalized on load.

**Comment by Thanood** [comment]
So, I've tried several `CommonChunkPlugin` variations and just failed miserably..
I still think this could be an answer or at least it seems like one can influence load order (common chunks first) and if you put all of jQuery/Materialize in there it should load first before aurelia.

I've tried with this `main.js`:

```javascript
export async function configure(aurelia) {
  await System.import('materialize-css');
  aurelia.use
    .standardConfiguration()
    .developmentLogging()
    .plugin('aurelia-materialize-bridge', bridge => bridge.useAll() );

  await aurelia.start();
  aurelia.setRoot('app');
}
```

My thought was  to separate out materialize and jQuery:

```javascript
const baseConfig = {
  entry: {
    'app': [/* this is filled by the aurelia-webpack-plugin */],
    'aurelia-bootstrap': coreBundles.bootstrap,
    'aurelia': coreBundles.aurelia.filter(pkg => coreBundles.bootstrap.indexOf(pkg) === -1),
    'materialize': ['materialize-css'],
    'jquery': ['jquery']
  },
  output: {
    path: outDir,
  }
}
```

Then configure a `CommonChunkPlugin`:

```javascript
var CommonsChunkPlugin = require("webpack/lib/optimize/CommonsChunkPlugin");
config.plugins = config.plugins.concat([
  new CommonsChunkPlugin({ name: 'commons.js', chunks: ['jquery', 'materialize'] })
]); 
```

(and several variations thereof)
But *this does not work at all*, I'm just writing it down for the record.

**Comment by Thanood** [comment]
About dimmed buttons, I managed to "solve" this by putting the following in `main.js` before aurelia starts:

```javascript
  if (Waves) {
    Waves.displayEffect = function() {};
  }
```

Not pretty, but it seems to work.

**Comment by Thanood** [comment]
Seems to be fixed with the solution given in #317.

**Comment by MaximBalaganskiy** [comment]
Sorry to spam here again but I've figured it out for CLI.

Waves are indeed initialised twice - once by materialize on every element click with this line
```
document.body.addEventListener('mousedown', showEffect, false);
```
And the second time by the bridge on `attached`.

I believe, the bridge should not call `Waves.attach()` because it's handled by `materialize`

**Comment by Thanood** [comment]
This is a strange thing because from the looks of it (event listener attached to `body`) it should work.
Unfortunately, this doesn't work in all cases.

We've had this for the AMD output and it broke some installations.
I'd need to look that up or try it again.
(I've also documented it somewhere but I couldn't find it)

**Comment by MaximBalaganskiy** [comment]
Can we only do the attach if materialise failed to do so? There must be some indication on the element that it's been enhanced

**Comment by Thanood** [comment]
Well, there's an event listener on the `body`. 😃 
The element itself is unchanged, though.

What about a setup option `.plugin('aurelia-materialize-bridge', b => b.all().preventWavesAttach()`?

**Comment by MaximBalaganskiy** [comment]
:) brilliant, that will work for everyone

**Comment by MaximBalaganskiy** [comment]
The offender here is webpack, not CLI :) CLI is actually working correctly.
`DOMContentLoaded` listener is added too late when webpack is used, hence the need to call `Waves.attach` in the attribute.

**Comment by MaximBalaganskiy** [comment]
Docs updated
